bpf/mb_process.c

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright © 2022 Merbridge Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#include "headers/helpers.h"
+#include "headers/maps.h"
+#include <linux/ptrace.h>
+
+struct xupid {
+    int nr;
+    long *ns;
+};
+
+/*
+for 5.7 and after
+struct pid
+{
+        refcount_t count;  // 4
+        unsigned int level; // 4
+        spinlock_t lock; // 4
+        // __pad 4 for align
+        struct hlist_head tasks[PIDTYPE_MAX]; // 8 * 4
+        struct hlist_head inodes; // 8
+        wait_queue_head_t wait_pidfd; // 24
+        struct rcu_head rcu; // 16
+        struct upid numbers[1]; // 16
+};
+
+#define XPID_OFFSET 88
+
+5.4-5.6
+struct pid
+{
+        refcount_t count; // 4
+        unsigned int level; // 4
+        struct hlist_head tasks[PIDTYPE_MAX]; // 8 * 4
+        wait_queue_head_t wait_pidfd; // 24
+        struct rcu_head rcu; // 16
+        struct upid numbers[1]; // 16
+};
+
+#define XPID_OFFSET 72
+
+*/
+
+#ifndef XPID_OFFSET
+#define XPID_OFFSET 88
+#endif
+
+struct xpid {
+    int __pad1;
+    int level;
+    char __pad2[XPID_OFFSET];
+    struct xupid numbers[1];
+};
+
+struct process_event {
+    int op; // 0 - fork, 1 - exit
+    int hostpid;
+    int level;
+    int levelpid;
```

Comment:
Levelpid refers to the process ID (Pid) at a particular level within a container environment that may contain multiple PID namespaces. For example, level = 0 refers to the corresponding pid on the host (hostpid).

This is primarily to support clusters created using tools like Kind. Since Merbridge runs inside a container and not in the host's PID Namespace, we need to read the Pid at the current level (Level pid).